### PR TITLE
Remove field refs from DatasetEditor

### PR DESCRIPTION
**Pull request by ranquild**
Depends on https://github.com/metabase/metabase/pull/45109
Related https://github.com/metabase/metabase/issues/42203
Related https://github.com/metabase/metabase/issues/25885
Related https://github.com/metabase/metabase/issues/33959
Related https://github.com/metabase/metabase/issues/36364

An attempt to simplify things by making metadata editing unaware of field refs. If this gets merged, we would rely on unique `column.name`. The PR doesn't fix anything on its own now.

Basically:
- `metadataDiff` becomes the single source of truth
- we use `column.name` for keys in `metadataDiff` and not field refs (which are not unique)
- we no longer change `result_metadata` in the current card before actually saving it
- we no longer try to make `metadataDiff` in sync with `result_metadata`. Instead, `metadataDiff` is applied on top of `result_metadata` in a selector. This was already the case, but because previously this code updated `result_metadata` at the same time, it had to deal with sync issues.

### CODE REVIEW

File: frontend/src/metabase-lib/v1/metadata/utils/models.ts [hunk 1]
```diff
@@ -148,42 +148,21 @@ export function getModelCacheSchemaName(databaseId: number, siteUUID: string) {
...
 
-  // always return metadata columns even if the corresponding viz settings don't exist
-  return columnMetadata
-    .map(column => ({
-      column,
-      index: getFieldIndexFromColumnVizSetting(column, columnSettings),
-    }))
-    .sort((a, b) => a.index - b.index)
-    .map(({ column }) => column);
+  return findColumnIndexesForColumnSettings(columns, columnSettings)
```
Comment (ranquild) [inline frontend/src/metabase-lib/v1/metadata/utils/models.ts:165]: Rely on the function used for viz settings in the main part of the product; it's much more sophisticated

File: frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx [hunk 2]
```diff
@@ -93,6 +94,7 @@ const TABLE_HEADER_HEIGHT = 45;
 function mapStateToProps(state) {
   return {
     metadata: getMetadata(state),
+    visualizationSettings: getVisualizationSettings(state),
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx:97]: Rely on the computed viz settings. This handles missing columns and backfilling `table.columns` when needed.

File: frontend/src/metabase-lib/v1/metadata/utils/models.ts [hunk 3]
```diff
@@ -3,14 +3,14 @@ import type Question from "metabase-lib/v1/Question";
 import type Database from "metabase-lib/v1/metadata/Database";
 import { getQuestionVirtualTableId } from "metabase-lib/v1/metadata/utils/saved-questions";
 import type NativeQuery from "metabase-lib/v1/queries/NativeQuery";
-import { isSameField } from "metabase-lib/v1/queries/utils/field-ref";
+import { findColumnIndexesForColumnSettings } from "metabase-lib/v1/queries/utils/dataset";
```
Comment (kamilmielnik) [inline frontend/src/metabase-lib/v1/metadata/utils/models.ts:6]: > Make sure you can change metadata for duplicated columns separately

It does not seem to work. It's broken in `master` as well but has slightly different symptoms.
Are you sure this PR can fix it?

https://github.com/metabase/metabase/assets/6830683/f1420a5a-9356-49ad-a151-cf9d96ca17b1

Comment (ranquild) [reply frontend/src/metabase-lib/v1/metadata/utils/models.ts:6]: @kamilmielnik you're correct - this PR is blocked until `column.name` is fixed and actually unique. Discussed here https://metaboat.slack.com/archives/C04CYTEL9N2/p1720016104124849

Comment (ranquild) [reply frontend/src/metabase-lib/v1/metadata/utils/models.ts:6]: Fixed now

File: e2e/test/scenarios/models/reproductions.cy.spec.ts [hunk 4]
```diff
@@ -668,3 +668,90 @@ describe("issue 39749", () => {
...
+      .clear()
+      .type(newName);
+  }
+
+  function verifyColumnName(name: string) {
+    tableHeaderClick(name);
+    cy.findByLabelText("Display name").should("have.value", name);
+  }
+
+  it("should allow to edit metadata for mbql models with self joins columns (metabase#25885)", () => {
```
Comment (ranquild) [inline e2e/test/scenarios/models/reproductions.cy.spec.ts:744]: Just to make it explicit, I'm not closing the issue as metadata reconciliation is broken on the BE right now.

File: frontend/src/metabase-lib/v1/metadata/utils/models.ts (hunk 3 above)
Comment (ranquild) [reply frontend/src/metabase-lib/v1/metadata/utils/models.ts:6]: Not fixed. I reverted this change as it becomes too much for one PR

File: src/metabase/query_processor/middleware/results_metadata.clj [hunk 5]
```diff
@@ -77,6 +78,8 @@
              (map? result)
              (update :data
                      assoc
+                     ;; TODO: We agreed on the name `:result_metadata` everywhere, and this needs updating.
+                     ;; It'll definitely break things on the FE, so a coordinated change is needed.
```
Comment (kamilmielnik) [inline src/metabase/query_processor/middleware/results_metadata.clj:82]: Let's make sure there's a GH issue for that

**Review by kamilmielnik — APPROVED** [review]

**Pull request merged**